### PR TITLE
Implement addBefore() & addAfter() methods of ColumnCollection

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description? | It implements `addBefore()` & `addAfter()` methods of ColumnCollection. These methods are especially useful when modifying grid's columns from hooks.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | n/a
| How to test?  | n/a

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9250)

### CODE REVIEW

File: src/Core/Grid/Column/ColumnCollection.php [hunk 1]
```diff
@@ -52,31 +51,19 @@ public function add(ColumnInterface $column)
      */
     public function addAfter($id, ColumnInterface $newColumn)
     {
-        if (!isset($this->items[$id])) {
-            throw new ColumnNotFoundException(sprintf(
-                'Column with id "%s" was not found.', $id
-            ));
-        }
+        $this->insertByPosition($id, $newColumn, 'after');
```
Comment (PierreRambaud) [inline src/Core/Grid/Column/ColumnCollection.php]: After and before can be const :)

Comment (sarjon) [reply src/Core/Grid/Column/ColumnCollection.php]: i would like to make them private constants, as they are not meant to be used outside this class, but it must support PHP 5.6. :/

Comment (PierreRambaud) [reply src/Core/Grid/Column/ColumnCollection.php]: If you don't want to use public const and allow only to value in `insertByPosition` you can use a boolean instead =)

Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: disagree on the boolean solution, we lose context :)

Even if these constants shouldn't be used, I'm in favor of constants (we may notify they should not be used outside of class with `@internal` annotation and then change public to private once we will be able to bump our PHP requirements.

File: src/Core/Grid/Column/ColumnCollection.php [hunk 2]
```diff
@@ -90,4 +77,34 @@ public function remove($id)
...
+    /**
+     * Insert new column into collection at given position
+     *
+     * @param string          $id        Existing column id
+     * @param ColumnInterface $newColumn Column to insert
+     * @param string          $position  Position: "before" or "after"
+     *
+     * @throws ColumnNotFoundException When column with gieven $id does not exist
+     */
+    private function insertByPosition($id, ColumnInterface $newColumn, $position)
```
Comment (mickaelandrieu) [inline src/Core/Grid/Column/ColumnCollection.php:100]: I'd like to see unit tests on that: they're easy to create and will cover one of the most used use cases of the Grid component.

**Review by mickaelandrieu — DISMISSED** [review]
Please provide some tests and go for public constants :)

File: src/Core/Grid/Column/ColumnCollection.php (hunk 2 above)
Comment (sarjon) [reply src/Core/Grid/Column/ColumnCollection.php:100]: will do. :)

Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php:100]: you'll save us a lot of technical debt doing this, believe me <3

File: tests/Unit/Core/Grid/Column/ColumnCollectionTest.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        return $columns;
+    }
+
+    /**
+     * @depends testItAddsColumnsBeforeGivenColumn
+     */
+    public function testItAddsColumnsAfterGivenColumn(ColumnCollection $columns)
+    {
+        $columns
+            ->addAfter('first', $this->createColumnMock('after_first'))
```
Comment (PierreRambaud) [inline tests/Unit/Core/Grid/Column/ColumnCollectionTest.php:82]: Do you have the time to add more tests like `->addBefore()->addAfter()->addBefore()->addAfter()` could be awesome :)

Comment (sarjon) [reply tests/Unit/Core/Grid/Column/ColumnCollectionTest.php:82]: sure. :)

**Comment by sarjon** [comment]
@mickaelandrieu done, ready for review. :)

File: src/Core/Grid/Column/ColumnCollection.php [hunk 4]
```diff
@@ -90,4 +87,34 @@ public function remove($id)
...
+        }
+
+        $existingColumnKeyPosition = array_search($id, array_keys($this->items));
+
+        if (self::POSITION_AFTER === $position) {
+            $existingColumnKeyPosition++;
+        }
+
+        $columns = array_slice($this->items, 0, $existingColumnKeyPosition);
+        $columns = array_merge($columns, [$newColumn->getId() => $newColumn]);
```
Comment (PierreRambaud) [inline src/Core/Grid/Column/ColumnCollection.php]: You can use only one array_merge here

File: src/Core/Grid/Column/ColumnCollection.php [hunk 5]
```diff
@@ -90,4 +87,34 @@ public function remove($id)
...
+        $columns = array_merge($columns, array_slice($this->items, $existingColumnKeyPosition));
```
Comment (PierreRambaud) [inline src/Core/Grid/Column/ColumnCollection.php]: You can improve performance by using this:
```php
$columns = array_slice($this->items, 0, $existingColumnKeyPosition, true) +
    [$newColumn->getId() => $newColumn] +
    array_slice($this->items, $existingColumnKeyPosition, count($this->items) - 1, true) ;
```

union `+` is eight faster than `array_merge`. More you forget to add true to preserved keys parameters
> Note that array_slice() will reorder and reset the numeric array indices by default. You can change this behaviour by setting

File: tests/Unit/Core/Grid/Column/ColumnCollectionTest.php [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            ->addBefore('first', $this->createColumnMock('before_first'))
+            ->addBefore('first', $this->createColumnMock('before_first_2'))
+            ->addAfter('first', $this->createColumnMock('after_first'))
+            ->addBefore('first', $this->createColumnMock('before_first_3'))
+            ->remove('non_existing')
+            ->addAfter('after_first', $this->createColumnMock('after_first_2'))
+            ->add($this->createColumnMock('second'))
+            ->remove('to_be_removed')
+            ->add($this->createColumnMock('third'))
+            ->addAfter('second', $this->createColumnMock('after_second'))
```
Comment (PierreRambaud) [inline tests/Unit/Core/Grid/Column/ColumnCollectionTest.php:128]: Now, can you test with other than string key, maybe numeric for example =)

Comment (mickaelandrieu) [reply tests/Unit/Core/Grid/Column/ColumnCollectionTest.php:128]: ping @sarjon this is not finished here ^^

Comment (sarjon) [reply tests/Unit/Core/Grid/Column/ColumnCollectionTest.php:128]: done.

**Comment by sarjon** [comment]
ping @PierreRambaud

**Review by mickaelandrieu — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks again @sarjon and @mickaelandrieu ;)

**Pull request merged**